readme.md
Object.observe polyfill
=======================

`Object.observe` polyfill based on [EcmaScript 7 spec](http://arv.github.io/ecmascript-object-observe/). Read the [documentation](doc/index.md) for more detailed informations. Check the [changelog](changelog.md) to see what's changed.

For a polyfill for `Array.observe`, you can find something [here](https://github.com/MaxArt2501/array-observe).

## Installation

This polyfill extends the native `Object` and doesn't have any dependencies, so loading it is pretty straightforward:

```html
<script src="object-observe.js"></script>
```

Using bower:

```bash
$ bower install object.observe
```

Or in node.js:

```bash
$ npm install object.observe
```

That's it. If the environment doesn't already support `Object.observe`, the shim is installed and ready to use.

For client side usage, if you only need the "light" version of the polyfill (see the [documentation](doc/index.md)), replace `object-observe.js` with `object-observe-lite.js`. You can also use the minified versions of the same files (replacing `.js` with `.min.js`). If you want to use the "light" version on the server side, you can either directly reference the desired version with `require` (as in `require("./node_modules/object.observe/dist/object-observe-lite.js");`) or change the `"main"` reference in [`package.json`](package.json).

For more details about loading the polyfill on a client, read the [documentation](doc/index.md#loading-on-a-client).

## Limitations and caveats

* Because properties are polled, when more than one change is made synchronously to the same property, it won't get caught. This means that this won't notify any event:

  ```js
  var object = { foo: null };
  Object.observe(object, function(changes) {
      console.log("Changes: ", changes);
  });

  object.foo = "bar";
  object.foo = null;
  ```

  `Object.prototype.watch` could help in this case, but it would be a partial solution.

* Property changes may not be reported in the correct order. The polyfill performs the checks - and issues the notifications - in this order:

  1. `"add"` and `"update"`, or `"reconfigure"`
  2. `"delete"` or `"reconfigure"`
  3. `"preventExtensions"`
  4. `"setPrototype"`

  This means that the `"add"` change is listed *before* the `"delete"` in the following snippet:

  ```js
  var obj = { foo: "bar" };
  Object.observe(foo, ...);
  delete obj.foo;
  obj.bar = "foo";
  ```

  Due to the nature of the shim, there's nothing that can be done about it.

* Environments that don't support [`Object.getOwnPropertyNames`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames) (most notably, Internet Explorer prior to version 9) can't detect changes to non-enumerable properties:

  ```js
  var re = /some regex/g;
  Object.observe(re, ...);

  re.lastIndex = 10;
  // Nothing happens...
  ```

  There's no way to prevent this limitation. Developers that need to support those environments should be careful about observing non-plain objects. A special polyfill is provided for arrays and the `length` property.

* Although the polyfill can work on DOM nodes or other *host* objects, the results may be impredictable. Moreover, in older environments like IE8-, observing nodes can be a cumbersome and memory hogging operation: they have a lot of enumerable properties that `Object.observe` should *not* check. Just don't observe nodes: it's not the point of `Object.observe`.

* **Possible memory leaks**: remember to `unobserve` the objects you want to be garbage collected. This can be avoided with native implementations of `Object.observe`, but due to the fact that in this polyfill observed objects are held in internal [maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), they can't be GC'ed until they're unobserved. ([`WeakMap`s](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) could solve this particular issue, but of course they're not for every environment - a shim for `Map` is used when not supported - and it's also not possible to iterate through their entries.)

* Dirty checking can be intensive. Memory occupation can grow to undesirable levels, not to mention CPU load. Pay attention to the number of objects your application needs to observe, and consider whether a polyfill is actually good for you.

## Browser support

This polyfill has been tested (and is working) in the following environments:

* Firefox 35-39 stable and 37-41 Developer Edition
* Internet Explorer 11
* Microsoft Edge 20
* Safari desktop 8
* Safari iOS 8.2
* Android browser 4.4
* Internet Explorer 5, 7, 8, 9, 10 (as IE11 in emulation mode)
* node.js 0.10.33-40

It also does *not* overwrite the native implentation in Chrome 36+, Opera 23+, node.js 0.11.13+ and io.js.

## Performances

Some benchmarks have been done on the dirty checking loop. See [the docs](doc/benchmarks.md) for more details.

## License

The MIT License (MIT)

Copyright (c) 2015 Massimo Artizzu (MaxArt2501)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


changelog.md
## v0.2.6

* Apply fix from MaxArt2501 at https://github.com/MaxArt2501/object-observe/pull/17 for #16

## v0.2.5

2015-10-13

* Fixes issue #16

## v0.2.4

2015-03-31

* Code linting
* Fixes issue #4

## v0.2.3

2015-03-06

* Fix on observing arrays in IE8-
* Removed DOM node detection (just don't observe nodes)
* Throwing on non-object accept lists
* Small code changes
* Unbundled utils, set as dev dependencies
* Added source maps for the minified versions

## v0.2.2

2015-02-11

* Uses `Object.getOwnPropertyNames` instead of `Object.keys` (issue #1)
* Functions can now be observed
* More tests

## v0.2.1

2015-01-28

* Desynch event delivering on `Object.unobserve`
* Package restructuring, splitted documentation
* `npm` module (`object.observe`)
* `bower` module (`object.observe`)

## v0.2.0

2015-01-26

* Code refactoring for correct event grouping
* Checks performed on a global loop with `requestAnimationFrame` (or a polyfilled version)
* Unused code cleanup
* Added some [mocha](http://mochajs.org/) tests

## v0.1.2

2015-01-21

* Support for `"reconfigure"` and `"setPrototype"` events
* New "light" version of the polyfill
* Various bugfixes
* Fixed links for specs

## v0.1.1

2015-01-19

* Fixed bugs in `Object.unobserve`
* Fixed bugs in `notifier.performChange`
* Used `Map`s when available - *much* cleaner code!
* Added some code documentation

## v0.1.0

2015-01-18 - First preliminary version

* `Object.observe`, `Object.unobserve`
* `Object.getNotifier`, `Object.deliverChangeRecords`
* No support for `"reconfigure"` and `"setPrototype"` events
* No `Array.observe`/`unobserve` yet


benchmark/benchmarks.js
(function(root, tests) {
    if (typeof define === "function" && define.amd)
        define(["expect", "object-observe"], tests);
    else if (typeof exports === "object")
        tests(require("benchmark"), global, require("../dist/object-observe.js"));
    else tests(root.Benchmark, root);
})(this, function(Benchmark, root) {
"use strict";

root.generateObject = function(numberOfProperties) {
    var object = {};

    for (var i = 0; i < numberOfProperties; i++)
        object["prop" + i] = i;

    return object;
}

var padspace = (new Array(81)).join(" ");
function padL(text, length) {
    return (padspace + text).slice(-length);
}
function padR(text, length) {
    return (text + padspace).slice(0, length);
}

var onDOM = typeof document !== "undefined";

if (onDOM) {
    var parent = document.getElementsByTagName("tbody")[0];

    var initBench = function() {
            var row = this.row = document.createElement("tr"),
                cell = document.createElement("td");
            cell.innerHTML = this.props;
            row.appendChild(cell);
            cell = document.createElement("td");
            cell.innerHTML = this.objects;
            row.appendChild(cell);
            cell = document.createElement("td");
            cell.className = "changes";
            cell.innerHTML = this.name;
            row.appendChild(cell);

            this.cells = [];
            cell = document.createElement("td");
            cell.className = "samples";
            this.cells.push(cell);
            row.appendChild(cell);
            cell = document.createElement("td");
            cell.className = "count";
            this.cells.push(cell);
            row.appendChild(cell);
            cell = document.createElement("td");
            cell.className = "frequency";
            this.cells.push(cell);
            row.appendChild(cell);

            parent.appendChild(row);
        },
        onCycle = function() {
            this.cells[0].innerHTML = ++this.samples;
            this.cells[1].innerHTML = this.count;
            this.cells[2].innerHTML = this.hz.toFixed(2);
        };
} else {
    var writeOut = typeof process === "undefined" || !process.stdout || !process.stdout.isTTY
            ? function(text) { console.log(text); }
            : function(text) { process.stdout.write(text); };
    writeOut("\x1b[1;37mProperties  Objects  Changes     Samples  Loops       FPS (Hz)\n");
    writeOut(          "---------------------------------------------------------------------");
    var initBench = function() {
            writeOut("\n\x1b[1;30m" + padL(this.props, 10) + "  " + padL(this.objects, 7) + "  " + padR(this.name, 48));
        },
        onCycle = function() {
            writeOut("\x1b[38D\x1b[1;31m" + padL(++this.samples, 9) + "\x1b[1;36m" + padL(this.count, 12)
                    + "\x1b[1;32m" + padL(this.hz.toFixed(2), 17) + "\x1b[0;37m");
        };
}

function errorBench(e) {
    console.log(e);
}

var benches = [];

function generateBenchGroup(props, objects) {
    var propsPerObject = props / objects;
    var options = {
        setup: "\
            var handler = function() {};\n\
            var objects = [];\n\
            for (var i = 0; i < " + objects + "; i++) {\n\
                objects[i] = generateObject(" + propsPerObject + ");\n\
                Object.observe(objects[i], handler);\n\
            }\n",
        teardown: "\
            for (var i = 0; i < " + objects + "; i++)\
                Object.unobserve(objects[i], handler);\n",
        props: props,
        objects: objects,
        onStart: initBench,
        onCycle: onCycle,
        onError: errorBench,
        onComplete: nextBench,
        async: true,
        samples: 0
    };
    var midIndex = objects >> 1,
        midProp = "prop" + (propsPerObject >> 1);

    benches.push(new Benchmark("none", "\
        Object.deliverChangeRecords(handler);\n", options));
    benches.push(new Benchmark("one", "\
        objects[" + midIndex + "]." + midProp + "++;\n\
        Object.deliverChangeRecords(handler);\n", options));
    benches.push(new Benchmark("all", "\
        for (var i = 0; i < " + objects + "; i++)\n\
            for (var j = 0; j < " + propsPerObject + "; j++)\n\
                objects[i][\"prop\" + j]++;\n\
        Object.deliverChangeRecords(handler);\n", options));
}

generateBenchGroup(10, 1);
generateBenchGroup(10, 2);
generateBenchGroup(100, 1);
generateBenchGroup(100, 20);
generateBenchGroup(1000, 1);
generateBenchGroup(1000, 200);

var index = 0;

function nextBench() {
    if (index >= benches.length) return;

    benches[index++].run();
}

nextBench();

});


object-observe-lite.js
/*!
 * Object.observe "lite" polyfill - v0.2.0
 * by Massimo Artizzu (MaxArt2501)
 * 
 * https://github.com/MaxArt2501/object-observe
 * 
 * Licensed under the MIT License
 * See LICENSE for details
 */

// Some type definitions
/**
 * This represents the data relative to an observed object
 * @typedef  {Object}                     ObjectData
 * @property {Map<Handler, HandlerData>}  handlers
 * @property {String[]}                   properties
 * @property {*[]}                        values
 * @property {Notifier}                   notifier
 */
/**
 * Function definition of a handler
 * @callback Handler
 * @param {ChangeRecord[]}                changes
*/
/**
 * This represents the data relative to an observed object and one of its
 * handlers
 * @typedef  {Object}                     HandlerData
 * @property {Map<Object, ObservedData>}  observed
 * @property {ChangeRecord[]}             changeRecords
 */
/**
 * @typedef  {Object}                     ObservedData
 * @property {String[]}                   acceptList
 * @property {ObjectData}                 data
*/
/**
 * Type definition for a change. Any other property can be added using
 * the notify() or performChange() methods of the notifier.
 * @typedef  {Object}                     ChangeRecord
 * @property {String}                     type
 * @property {Object}                     object
 * @property {String}                     [name]
 * @property {*}                          [oldValue]
 * @property {Number}                     [index]
 */
/**
 * Type definition for a notifier (what Object.getNotifier returns)
 * @typedef  {Object}                     Notifier
 * @property {Function}                   notify
 * @property {Function}                   performChange
 */
/**
 * Function called with Notifier.performChange. It may optionally return a
 * ChangeRecord that gets automatically notified, but `type` and `object`
 * properties are overridden.
 * @callback Performer
 * @returns {ChangeRecord|undefined}
 */

Object.observe || (function(O, A, root) {
    "use strict";

        /**
         * Relates observed objects and their data
         * @type {Map<Object, ObjectData}
         */
    var observed,
        /**
         * List of handlers and their data
         * @type {Map<Handler, Map<Object, HandlerData>>}
         */
        handlers,

        defaultAcceptList = [ "add", "update", "delete", "reconfigure", "setPrototype", "preventExtensions" ];

    // Functions for internal usage

        /**
         * Checks if the argument is a Node object. Uses duck typing if Node is
         * not defined.
         * @function isNode
         * @param {?*} node
         * @returns {Boolean}
         */
    var isNode = root.Node ? function(node) {
            return node && node instanceof root.Node;
        } : function(node) {
            // Duck typing
            return node && typeof node === "object"
                    && typeof node.nodeType === "number"
                    && typeof node.nodeName === "string";
        },

        /**
         * Checks if the argument is an Array object. Polyfills Array.isArray.
         * @function isArray
         * @param {?*} object
         * @returns {Boolean}
         */
        isArray = A.isArray || (function(toString) {
            return function (object) { return toString.call(object) === "[object Array]"; };
        })(O.prototype.toString),

        /**
         * Returns the index of an item in a collection, or -1 if not found.
         * Uses Array.prototype.indexOf is available.
         * @function inArray
         * @param {*} pivot           Item to look for
         * @param {Array} array
         * @param {Number} [start=0]  Index to start from
         * @returns {Number}
         */
        inArray = A.prototype.indexOf ? function(pivot, array, start) {
            return array.indexOf(pivot, start);
        } : function(pivot, array, start) {
            for (var i = start || 0; i < array.length; i++)
                if (array[i] === pivot)
                    return i;
            return -1;
        },

        /**
         * Returns an instance of Map, or a Map-like object is Map is not
         * supported or doesn't support forEach()
         * @function createMap
         * @returns {Map}
         */
        createMap = typeof root.Map === "undefined" || !Map.prototype.forEach ? function() {
            // Lightweight shim of Map. Lacks clear(), entries(), keys() and
            // values() (the last 3 not supported by IE11, so can't use them),
            // it doesn't handle the constructor's argument (like IE11) and of
            // course it doesn't support for...of.
            // Chrome 31-35 and Firefox 13-24 have a basic support of Map, but
            // they lack forEach(), so their native implementation is bad for
            // this polyfill. (Chrome 36+ supports Object.observe.)
            var keys = [], values = [];

            return {
                size: 0,
                has: function(key) { return inArray(key, keys) > -1; },
                get: function(key) { return values[inArray(key, keys)]; },
                set: function(key, value) {
                    var i = inArray(key, keys);
                    if (i === -1) {
                        keys.push(key);
                        values.push(value);
                        this.size++;
                    } else values[i] = value;
                },
                "delete": function(key) {
                    var i = inArray(key, keys);
                    if (i > -1) {
                        keys.splice(i, 1);
                        values.splice(i, 1);
                        this.size--;
                    }
                },
                forEach: function(callback/*, thisObj*/) {
                    for (var i = 0; i < keys.length; i++)
                        callback.call(arguments[1], values[i], keys[i], this);
                }
            };
        } : function() { return new Map(); },

        /**
         * Simple shim for Object.keys is not available
         * Misses checks on object, don't use as a replacement of Object.keys
         * @function getKeys
         * @param {Object} object
         * @returns {String[]}
         */
        getKeys = O.keys || function(object) {
            var keys = [], prop;
            for (prop in object)
                if (object.hasOwnProperty(prop))
                    keys.push(prop);
            return keys;
        },

        /**
         * Sets up the next check and delivering iteration, using
         * requestAnimationFrame or a (close) polyfill.
         * @function nextFrame
         * @param {function} func
         * @returns {number}
         */
        nextFrame = root.requestAnimationFrame || root.webkitRequestAnimationFrame || (function() {
            var initial = +new Date,
                last = initial;
            return function(func) {
                var now = +new Date;
                return setTimeout(function() {
                    func((last = +new Date) - initial);
                }, 17);
            };
        })(),

        /**
         * Sets up the observation of an object
         * @function doObserve
         * @param {Object} object
         * @param {Handler} handler
         * @param {String[]} [acceptList]
         */
        doObserve = function(object, handler, acceptList) {

            var data = observed.get(object);

            if (data)
                setHandler(object, data, handler, acceptList);
            else {
                data = createObjectData(object);
                setHandler(object, data, handler, acceptList);
                
                if (observed.size === 1)
                    // Let the observation begin!
                    nextFrame(runGlobalLoop);
            }
        },

        /**
         * Creates the initial data for an observed object
         * @function createObjectData
         * @param {Object} object
         */
        createObjectData = function(object, data) {
            var props = isNode(object) ? [] : getKeys(object),
                values = [], descs, i = 0,
                data = {
                    handlers: createMap(),
                    properties: props,
                    values: values,
                    notifier: retrieveNotifier(object, data)
                };

            while (i < props.length)
                values[i] = object[props[i++]];

            observed.set(object, data);

            return data;
        },

        /**
         * Performs basic property value change checks on an observed object
         * @function performPropertyChecks
         * @param {ObjectData} data
         * @param {Object} object
         * @param {String} [except]  Doesn't deliver the changes to the
         *                           handlers that accept this type
         */
        performPropertyChecks = function(data, object, except) {
            if (!data.handlers.size) return;

            var props, proplen, keys,
                values = data.values,
                descs = data.descriptors,
                i = 0, idx,
                key, value, ovalue,
                proto, descr;

            props = data.properties.slice();
            proplen = props.length;
            keys = getKeys(object);

            // Check for value additions/changes
            while (i < keys.length) {
                key = keys[i++];
                idx = inArray(key, props);
                value = object[key];

                if (idx === -1) {
                    addChangeRecord(object, data, {
                        name: key,
                        type: "add",
                        object: object
                    }, except);
                    data.properties.push(key);
                    values.push(value);
                } else {
                    ovalue = values[idx];
                    props[idx] = null;
                    proplen--;
                    if (ovalue === value ? ovalue === 0 && 1/ovalue !== 1/value 
                            : ovalue === ovalue || value === value) {
                        addChangeRecord(object, data, {
                            name: key,
                            type: "update",
                            object: object,
                            oldValue: ovalue
                        }, except);
                        data.values[idx] = value;
                    }
                }
            }

            // Checks if some property has been deleted
            for (i = props.length; proplen && i--;)
                if (props[i] !== null) {
                    addChangeRecord(object, data, {
                        name: props[i],
                        type: "delete",
                        object: object,
                        oldValue: values[i]
                    }, except);
                    data.properties.splice(i, 1);
                    data.values.splice(i, 1);
                    proplen--;
                }
        },

        /**
         * Sets up the main loop for object observation and change notification
         * It stops if no object is observed.
         * @function runGlobalLoop
         */
        runGlobalLoop = function() {
            if (observed.size) {
                observed.forEach(performPropertyChecks);
                handlers.forEach(deliverHandlerRecords);
                nextFrame(runGlobalLoop);
            }
        },

        /**
         * Deliver the change records relative to a certain handler, and resets
         * the record list.
         * @param {HandlerData} hdata
         * @param {Handler} handler
         */
        deliverHandlerRecords = function(hdata, handler) {
            if (hdata.changeRecords.length) {
                handler(hdata.changeRecords);
                hdata.changeRecords = [];
            }
        },

        /**
         * Returns the notifier for an object - whether it's observed or not
         * @function retrieveNotifier
         * @param {Object} object
         * @param {ObjectData} [data]
         * @returns {Notifier}
         */
        retrieveNotifier = function(object, data) {
            if (arguments.length < 2)
                data = observed.get(object);

            /** @type {Notifier} */
            return data && data.notifier || {
                /**
                 * @method notify
                 * @see http://arv.github.io/ecmascript-object-observe/#notifierprototype._notify
                 * @memberof Notifier
                 * @param {ChangeRecord} changeRecord
                 */
                notify: function(changeRecord) {
                    changeRecord.type; // Just to check the property is there...

                    // If there's no data, the object has been unobserved
                    var data = observed.get(object);
                    if (data) {
                        var recordCopy = { object: object }, prop;
                        for (prop in changeRecord)
                            if (prop !== "object")
                                recordCopy[prop] = changeRecord[prop];
                        addChangeRecord(object, data, recordCopy);
                    }
                },

                /**
                 * @method performChange
                 * @see http://arv.github.io/ecmascript-object-observe/#notifierprototype_.performchange
                 * @memberof Notifier
                 * @param {String} changeType
                 * @param {Performer} func     The task performer
                 * @param {*} [thisObj]        Used to set `this` when calling func
                 */
                performChange: function(changeType, func/*, thisObj*/) {
                    if (typeof changeType !== "string")
                        throw new TypeError("Invalid non-string changeType");

                    if (typeof func !== "function")
                        throw new TypeError("Cannot perform non-function");

                    // If there's no data, the object has been unobserved
                    var data = observed.get(object),
                        prop, changeRecord,
                        result = func.call(arguments[2]);

                    data && performPropertyChecks(data, object, changeType);

                    // If there's no data, the object has been unobserved
                    if (data && result && typeof result === "object") {
                        changeRecord = { object: object, type: changeType };
                        for (prop in result)
                            if (prop !== "object" && prop !== "type")
                                changeRecord[prop] = result[prop];
                        addChangeRecord(object, data, changeRecord);
                    }
                }
            };
        },

        /**
         * Register (or redefines) an handler in the collection for a given
         * object and a given type accept list.
         * @function setHandler
         * @param {Object} object
         * @param {ObjectData} data
         * @param {Handler} handler
         * @param {String[]} acceptList
         */
        setHandler = function(object, data, handler, acceptList) {
            var hdata = handlers.get(handler), odata;
            if (!hdata)
                handlers.set(handler, hdata = {
                    observed: createMap(),
                    changeRecords: []
                });
            hdata.observed.set(object, {
                acceptList: acceptList.slice(),
                data: data
            });
            data.handlers.set(handler, hdata);
        },

        /**
         * Adds a change record in a given ObjectData
         * @function addChangeRecord
         * @param {Object} object
         * @param {ObjectData} data
         * @param {ChangeRecord} changeRecord
         * @param {String} [except]
         */
        addChangeRecord = function(object, data, changeRecord, except) {
            data.handlers.forEach(function(hdata) {
                var acceptList = hdata.observed.get(object).acceptList;
                // If except is defined, Notifier.performChange has been
                // called, with except as the type.
                // All the handlers that accepts that type are skipped.
                if ((typeof except !== "string"
                        || inArray(except, acceptList) === -1)
                        && inArray(changeRecord.type, acceptList) > -1)
                    hdata.changeRecords.push(changeRecord);
            });
        };

    observed = createMap();
    handlers = createMap();

    /**
     * @function Object.observe
     * @see http://arv.github.io/ecmascript-object-observe/#Object.observe
     * @param {Object} object
     * @param {Handler} handler
     * @param {String[]} [acceptList]
     * @throws {TypeError}
     * @returns {Object}               The observed object
     */
    O.observe = function observe(object, handler, acceptList) {
        if (object === null || typeof object !== "object")
            throw new TypeError("Object.observe cannot observe non-object");

        if (typeof handler !== "function")
            throw new TypeError("Object.observe cannot deliver to non-function");

        if (O.isFrozen && O.isFrozen(handler))
            throw new TypeError("Object.observe cannot deliver to a frozen function object");

        if (!isArray(acceptList))
            acceptList = defaultAcceptList;

        doObserve(object, handler, acceptList);

        return object;
    };

    /**
     * @function Object.unobserve
     * @see http://arv.github.io/ecmascript-object-observe/#Object.unobserve
     * @param {Object} object
     * @param {Handler} handler
     * @throws {TypeError}
     * @returns {Object}         The given object
     */
    O.unobserve = function unobserve(object, handler) {
        if (object === null || typeof object !== "object")
            throw new TypeError("Object.unobserve cannot unobserve non-object");

        if (typeof handler !== "function")
            throw new TypeError("Object.unobserve cannot deliver to non-function");

        var hdata = handlers.get(handler), odata;

        if (hdata && (odata = hdata.observed.get(object))) {
            hdata.observed.forEach(function(odata, object) {
                performPropertyChecks(odata.data, object);
            });
            deliverHandlerRecords(hdata, handler);

            // In Firefox 13-18, size is a function, but createMap should fall
            // back to the shim for those versions
            if (hdata.observed.size === 1 && hdata.observed.has(object))
                handlers["delete"](handler);
            else hdata.observed["delete"](object);

            if (odata.data.handlers.size === 1)
                observed["delete"](object);
            else odata.data.handlers["delete"](handler);
        }

        return object;
    };

    /**
     * @function Object.getNotifier
     * @see http://arv.github.io/ecmascript-object-observe/#GetNotifier
     * @param {Object} object
     * @throws {TypeError}
     * @returns {Notifier}
     */
    O.getNotifier = function getNotifier(object) {
        if (object === null || typeof object !== "object")
            throw new TypeError("Object.getNotifier cannot getNotifier non-object");

        if (O.isFrozen && O.isFrozen(object)) return null;

        return retrieveNotifier(object);
    };

    /**
     * @function Object.deliverChangeRecords
     * @see http://arv.github.io/ecmascript-object-observe/#Object.deliverChangeRecords
     * @see http://arv.github.io/ecmascript-object-observe/#DeliverChangeRecords
     * @param {Handler} handler
     * @throws {TypeError}
     */
    O.deliverChangeRecords = function deliverChangeRecords(handler) {
        if (typeof handler !== "function")
            throw new TypeError("Object.deliverChangeRecords cannot deliver to non-function");

        var hdata = handlers.get(handler);
        if (hdata) {
            hdata.observed.forEach(function(odata, object) {
                performPropertyChecks(odata.data, object);
            });
            deliverHandlerRecords(hdata, handler);
        }
    };

})(Object, Array, this);